### ISSUE TITLE
[dataclasses] I3MCTree read/write machine dependence (Trac #1503)

### ISSUE THREAD

**Issue by icecube-trac** [icecube-trac/tix2#1495]
Writing a tree on a 32-bit Ubuntu machine and reading on a 64-bit Ubuntu machine fails with the following message :

In [5]: frame = f.pop_frame()

In [6]: print frame
[ I3Frame  (DAQ):
  'I3EventHeader' [DAQ] ==> I3EventHeader (99)
  'I3MCTree' [DAQ] ==> TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> > (418)
  'I3MCWeightDict' [DAQ] ==> I3Map<string, double> (1110)
  'MCTimeIncEventID' [DAQ] ==> I3PODHolder<bool> (27)
  'NuGPrimary' [DAQ] ==> I3Particle (150)
]

In [7]: mctree = frame["I3MCTree"]
WARN (dataclasses): unknown RDMC code "-14" cannot be converted to a I3Particle::ParticleType. It will appear as "unknown". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])
WARN (dataclasses): unknown RDMC code "65535" cannot be converted to a I3Particle::ParticleType. It will appear as "unknown". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])
WARN (dataclasses): unknown RDMC code "2001098" cannot be converted to a I3Particle::ParticleType. It will appear as "unknown". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])
python: /usr/include/boost/dynamic_bitset/dynamic_bitset.hpp:616: boost::dynamic_bitset<Block, Allocator>::~dynamic_bitset() [with Block = long unsigned int; Allocator = std::allocator<long unsigned int>]: Assertion `m_check_invariants()' failed.
[1]    7057 abort (core dumped)  ipython

Other frame objects deserialize just fine.

Both machines are 14.04 LTS built against system packages.  My guess is this is a 32/64 bit issue.

Migrated from https://code.icecube.wisc.edu/ticket/1503
```json
{
    "status": "closed", 
    "changetime": "2016-03-18T21:14:15", 
    "description": "Writing a tree on a 32-bit Ubuntu machine and reading on a 64-bit Ubuntu machine fails with the following message :\n\nIn [5]: frame = f.pop_frame()\n\nIn [6]: print frame\n[ I3Frame  (DAQ):\n  'I3EventHeader' [DAQ] ==> I3EventHeader (99)\n  'I3MCTree' [DAQ] ==> TreeBase::Tree<I3Particle, I3ParticleID, __gnu_cxx::hash<I3ParticleID> > (418)\n  'I3MCWeightDict' [DAQ] ==> I3Map<string, double> (1110)\n  'MCTimeIncEventID' [DAQ] ==> I3PODHolder<bool> (27)\n  'NuGPrimary' [DAQ] ==> I3Particle (150)\n]\n\n\nIn [7]: mctree = frame[\"I3MCTree\"]\nWARN (dataclasses): unknown RDMC code \"-14\" cannot be converted to a I3Particle::ParticleType. It will appear as \"unknown\". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])\nWARN (dataclasses): unknown RDMC code \"65535\" cannot be converted to a I3Particle::ParticleType. It will appear as \"unknown\". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])\nWARN (dataclasses): unknown RDMC code \"2001098\" cannot be converted to a I3Particle::ParticleType. It will appear as \"unknown\". (I3Particle.cxx:864 in void I3Particle::load(Archive&, unsigned int) [with Archive = boost::archive::portable_binary_iarchive])\npython: /usr/include/boost/dynamic_bitset/dynamic_bitset.hpp:616: boost::dynamic_bitset<Block, Allocator>::~dynamic_bitset() [with Block = long unsigned int; Allocator = std::allocator<long unsigned int>]: Assertion `m_check_invariants()' failed.\n[1]    7057 abort (core dumped)  ipython\n\n\nOther frame objects deserialize just fine.\n\nBoth machines are 14.04 LTS built against system packages.  My guess is this is a 32/64 bit issue.", 
    "reporter": "olivas", 
    "cc": "", 
    "resolution": "fixed", 
    "_ts": "1458335655846260", 
    "component": "combo core", 
    "summary": "[dataclasses] I3MCTree read/write machine dependence", 
    "priority": "major", 
    "keywords": "", 
    "time": "2016-01-11T03:06:13", 
    "milestone": "", 
    "owner": "david.schultz", 
    "type": "defect"
}
```

**Comment by icecube-trac** [comment]
Trac update at 20160111T16:24:10: david.schultz commented:

> Who wants to bet it's the `long unsigned int` size that's different?
> 
> Found the reference to why this is.  `boost::dynamic_bitset::Block` is by default `unsigned long`, so I guess I just took that without thinking.  Maybe they should change their default?
> 
> As for a fix, since all production computing is 64bit I should be able to maintain backwards compatibility by forcing this to 8 bytes.  32bit files will be unreadable, but they already are I guess.

**Comment by icecube-trac** [comment]
Trac update at 20160111T16:24:23: david.schultz changed status from "new" to "accepted"

**Comment by icecube-trac** [comment]
Trac update at 20160111T16:27:49: olivas commented:

> Yeah.  I can't imagine we have any 32 bit systems in production, so likely not a huge deal.  We should have a note somewhere in the docs to point people to if they ever run into this problem.  Maybe a good time to start an "IceCube Software FAQ" or "Common/Possible Problems and Pitfalls" or "Knowledge Database" or something...

**Comment by icecube-trac** [comment]
Trac update at 20160111T16:48:39: 
* david.schultz changed status from "accepted" to "closed"
* david.schultz changed resolution from "" to "fixed"
* david.schultz commented:

> In changeset 140847:
> 
> ```text
> #!CommitTicketReference repository="IceCube" revision="140847"
> use a fixed size datatype when serializing. this should be transparent to 64bit machines, but 32bit machines will not be backwards-compatible. closes #1495
> ```

**Comment by icecube-trac** [comment]
Trac update at 20160318T21:14:15: 
* olivas commented:

> Milestone February 2016 deleted


* olivas changed milestone from "February 2016" to ""
